grease/src/Grease/Shape.hs

```diff
@@ -314,3 +340,260 @@ deriving instance (ShowExt ext tag, ShowF tag) => Show (ArgShapes ext tag tys)
+    let mp = MDwarf.subTypeMap sprog
+    (mbtypeApp, _) <- Map.lookup vrTy mp
+    rightToMaybe mbtypeApp
+
+constructPtrTarget :: HasPtrWidth w => Subprogram -> MDwarf.TypeApp -> StateT VisitState Maybe (PtrTarget w NoTag)
+constructPtrTarget sprog tyApp =
+  PtrTarget Nothing <$> shapeSeq tyApp
+ where
+  -- TODO: are these always byte fields?
+  ishape w = lift $ Just $ Seq.singleton $ Initialized NoTag (toBytes w)
+  padding :: (Integral a) => a -> MemShape w NoTag
+  padding w = Uninitialized (toBytes w)
+  -- if we dont know where to place members we have to fail/just place some bytes
+  buildMember :: HasPtrWidth w => Word64 -> MDwarf.Member -> StateT VisitState Maybe (Word64, Seq.Seq (MemShape w NoTag))
+  buildMember loc mem =
+    let memRes =
+          ( \memLoc ->
+              do
+                padd <- lift $ Just (if memLoc == loc then Seq.empty else Seq.singleton (padding $ memLoc - loc))
+                membershape <- shapeOfTyApp $ MDwarf.memberType mem
+                let memByteSize = sum $ memShapeSize ?ptrWidth <$> membershape
+                let nextLoc = memLoc + fromIntegral memByteSize
+                lift $ Just (nextLoc, padd Seq.>< membershape)
+          )
+            =<< (lift $ MDwarf.memberLoc mem)
+     in memRes
+  shapeOfTyApp :: (HasPtrWidth w) => MDwarf.TypeRef -> StateT VisitState Maybe (Seq.Seq (MemShape w NoTag))
+  shapeOfTyApp x = shapeSeq =<< trace ("extracting type " ++ (show $ extractType sprog x)) (lift . extractType sprog) =<< pure x
+
+  shapeSeq :: (HasPtrWidth w) => MDwarf.TypeApp -> StateT VisitState Maybe (Seq.Seq (MemShape w NoTag))
+  shapeSeq (MDwarf.UnsignedIntType w) = ishape w
+  shapeSeq (MDwarf.SignedIntType w) = ishape w
+  shapeSeq MDwarf.SignedCharType = ishape (1 :: Int)
+  shapeSeq MDwarf.UnsignedCharType = ishape (1 :: Int)
+  -- Need modification to DWARF to collect counts properly + evaluate ops for ub
+  shapeSeq (MDwarf.ArrayType _elTy _ub) = lift $ Nothing
+  -- need to compute padding between els, infront of els and at end
+  shapeSeq (MDwarf.StructType sdecl) =
+    let structSize = MDwarf.structByteSize sdecl
+     in trace
+          "in struct"
+          ( do
+              (currLoc, seqs) <-
+                foldM
+                  ( \(currloc, seqs) mem ->
+                      do
+                        (nextLoc, additionalShapes) <- buildMember currloc mem
+                        lift $ Just (nextLoc, seqs Seq.>< additionalShapes)
+                  )
+                  (0, Seq.empty)
+                  (MDwarf.structMembers sdecl)
+              endPad <- lift $ Just (if currLoc >= structSize then Seq.empty else Seq.singleton (padding $ structSize - currLoc))
+              lift $ Just $ (seqs Seq.>< endPad)
+          )
+  shapeSeq (MDwarf.PointerType _ maybeRef) =
+    let mshape = constructPtrMemShapeFromRef sprog =<< (lift $ maybeRef)
+     in Seq.singleton <$> mshape
+  shapeSeq _ = lift $ Nothing
+
+type VisitState = Map.Map MDwarf.TypeRef Int
```

Comment:
FWIW I personally tend to avoid type synonyms - it's just one more thing to keep track of when reading the code. In a situation like this, I'd probably make this a `newtype` and add dedicated helpers like `visitType :: MDwarf.TypeRef -> StateT VisitState m ()`